File: 3530.maximum-profit-from-valid-topological-order-in-dag.cpp
#
# @lc app=leetcode id=3530 lang=cpp
#
# [3530] Maximum Profit from Valid Topological Order in DAG
#
# @lc code=start
class Solution {
public:
    int maxProfit(int n, vector<vector<int>>& edges, vector<int>& score) {
        // Build adjacency list for incoming edges (prerequisites)
        vector<vector<int>> incoming(n);
        for (auto& edge : edges) {
            incoming[edge[1]].push_back(edge[0]);
        }
        
        // DP with bitmask: dp[mask] = max profit when nodes in mask are processed
        vector<int> dp(1 << n, -1);
        dp[0] = 0;
        
        for (int mask = 0; mask < (1 << n); mask++) {
            if (dp[mask] == -1) continue;
            
            int position = __builtin_popcount(mask) + 1;
            
            // Try adding each unprocessed node
            for (int node = 0; node < n; node++) {
                if (mask & (1 << node)) continue; // already processed
                
                // Check if all prerequisites are satisfied
                bool canAdd = true;
                for (int prereq : incoming[node]) {
                    if (!(mask & (1 << prereq))) {
                        canAdd = false;
                        break;
                    }
                }
                
                if (canAdd) {
                    int newMask = mask | (1 << node);
                    int newProfit = dp[mask] + score[node] * position;
                    dp[newMask] = max(dp[newMask], newProfit);
                }
            }
        }
        
        return dp[(1 << n) - 1];
    }
};
# @lc code=end

File: 3530.maximum-profit-from-valid-topological-order-in-dag.java
#
# @lc app=leetcode id=3530 lang=java
#
# [3530] Maximum Profit from Valid Topological Order in DAG
#
# @lc code=start
class Solution {
    public int maxProfit(int n, int[][] edges, int[] score) {
        int[] deps = new int[n]; // dependencies as bitmask
        for (int[] e : edges) {
            int u = e[0], v = e[1];
            deps[v] |= (1 << u);
        }
        int N = 1 << n;
        int[] dp = new int[N];
        for (int mask = 0; mask < N; ++mask) {
            int pos = Integer.bitCount(mask); // number of processed nodes
            for (int u = 0; u < n; ++u) {
                if ((mask & (1 << u)) == 0 && (deps[u] & ~mask) == 0) {
                    int newMask = mask | (1 << u);
                    dp[newMask] = Math.max(dp[newMask], dp[mask] + score[u] * (pos + 1));
                }
            }
        }
        return dp[N-1];
    }
}
# @lc code=end